pytorch_lightning/callbacks/quantization.py

```diff
@@ -0,0 +1,151 @@
+                - custom lambda function with single trainer argument,
+                    see example when you limit call only for last epoch::
+
+                    def custom_trigger_last(trainer):
+                        return trainer.current_epoch == (trainer.max_epochs - 1)
+
+                    QuantizationAwareTraining(lambda_trigger=custom_trigger_last)
+
+            modules_to_fuse: allows you fuse a few layer together as you can see in `diagram
+             <https://pytorch.org/docs/stable/quantization.html#quantization-aware-training>_`
+                to find what layer types can be fue=sed check https://github.com/pytorch/pytorch/pull/43286
+        """
+        if not isinstance(qconfig, (str, QConfig)):
+            raise MisconfigurationException(f"Unsupported qconfig: f{self._qconfig}")
+        self._qconfig = qconfig
+        self._lambda_trigger = lambda_trigger
+        self.modules_to_fuse = modules_to_fuse
+        self._forward_calls = 0
+
+    def increment_forward(self):
+        self._forward_calls += 1
+
+    def _check_feasible_fuse(self, model):
+        if not self.modules_to_fuse:
+            return True
+        for group in self.modules_to_fuse:
+            if not all(_recursive_hasattr(model, m) for m in group):
+                raise MisconfigurationException(
+                    f'You have requested to fuse {group} but one or more of them is not your model attributes'
+                )
+        return True
+
+    def on_fit_start(self, trainer, pl_module):
+        # QuantStub converts tensors from floating point to quantized
+        pl_module.quant = torch.quantization.QuantStub()
+        # DeQuantStub converts tensors from quantized to floating point
+        pl_module.dequant = torch.quantization.DeQuantStub()
+        # manually specify where tensors will be converted from quantized
+        # to floating point in the quantized model
+        pl_module.forward = wrap_quantize_forward_context(
+            quant_cb=self, model=pl_module, func=pl_module.forward, trigger_func=self._lambda_trigger
+        )
+
+        # attach a global qconfig, which contains information about what kind
+        # of observers to attach. Use 'fbgemm' for server inference
+        if isinstance(self._qconfig, str):
+            pl_module.qconfig = torch.quantization.get_default_qat_qconfig(self._qconfig)
+        elif isinstance(self._qconfig, QConfig):
+            pl_module.qconfig = self._qconfig
+
+        if self.modules_to_fuse:
+            self._check_feasible_fuse(pl_module)
+            torch.quantization.fuse_modules(pl_module, self.modules_to_fuse, inplace=True)
+
+        # Prepare the model for QAT. This inserts observers and fake_quants in
+        # the model that will observe weight and activation tensors during calibration.
+        torch.quantization.prepare_qat(pl_module, inplace=True)
+
+    def on_fit_end(self, trainer, pl_module):
```

Comment:
model should be cleaned out of extra modules and wrapper no ?

Reply:
that is a good question, if you clean it you can pass to the model only quant inputs, if we keep these layer quant/dequant you can keep reading anything as before and still hold all other layers quantized -> small

Reply:
Let's keep it then. Can you still torch-script the model with the wrapper ?

Reply:
I have cleaned it so let's see :]

Reply:
I am curious if you could do it with hook even if you in fact insert an extra layer on-the-fly which is missing in the model init declaration